core/src/main/scala/filodb.core/reprojector/FiloMemTable.scala

```diff
@@ -0,0 +1,216 @@
+      }
+    }
+  }
+
+  // Converts rows to chunks, merging with previous chunks as needed, and invokes callbacks
+  // Also updates the partKeyMap which maps every sort key to its chunk ID and row index
+  private def flushRowsToChunks(): Unit = synchronized {
+    // VERY IMPORTANT: pass false to cancel so this task is not interrupted
+    // Cancel any other scheduled tasks, if one was scheduled
+    flushTask.foreach(_.cancel(false))
+    builder.reset()
+
+    // Last chunk written partial?  Get chunks out, populate builder, remove from chunks
+    if (readers.nonEmpty && readers.last.parsers(0).length < chunkSize) {
+      val reader = readers.last
+      for { i <- 0 until reader.parsers(0).length optimized } {
+        reader.rowNo = i
+        builder.addRow(reader)
+      }
+      chunks.remove(chunks.length - 1, 1)
+      readers.remove(readers.length - 1, 1)
+    }
+
+    // Add new rows to builder, adding to partKeyMap in the meantime
+    val baseLength = builder.builders.head.length   // Nonzero only if partial chunks added
+    val rowsToAdd = Math.min(chunkSize - baseLength, tempRows.length)
+    val nextChunkIndex = chunks.length
+    val sortKeyFunc = projection.sortKeyFunc
+
+    for { i <- 0 until rowsToAdd optimized } {
+      val row = tempRows(i)
+      builder.addRow(row)
+      val keyMap = getKeyMap(projection.partitionFunc(row))
+      keyMap.put(sortKeyFunc(row), chunkRowIdToLong(nextChunkIndex, baseLength + i))
+    }
+
+    // Add chunks
+    val chunkAray = new Array[ByteBuffer](filoSchema.length)
+    builder.convertToBytes().foreach { case (colName, bb) => chunkAray(colIdToNumber(colName)) = bb }
+    chunks += chunkAray
+    readers += new FastFiloRowReader(chunkAray, clazzes)
+
+    invokeAndCleanupCallbacks(rowsToAdd)
+    flushTask = None
+  }
+
+  def close(): Unit = {}
+
+  /**
+   * === Row ingest, read, delete operations ===
+   */
+  def ingestRows(rows: Seq[RowReader])(callback: => Unit): Unit = if (rows.nonEmpty) {
+    tempRows.synchronized {
+      val oldLen = tempRows.length
+      tempRows ++= rows
+      callbacks += ((oldLen, tempRows.length - 1, { () => callback }))
+    }
+    while (tempRows.length >= chunkSize) flushRowsToChunks()
+    // If rows have not been flushed, schedule a task in the future to flush it
+    if (tempRows.nonEmpty && !flushTask.isDefined) {
+      val fut = scheduler.schedule(new Runnable { def run: Unit = { flushRowsToChunks() }},
+                                   flushInterval.toMillis,
+                                   TimeUnit.MILLISECONDS)
+      flushTask = Some(fut)
+    }
+  }
+
+  /**
+   * Forces all the rows in tempRows to be committed into the chunk store and disk storage / WAL.
+   * This might be necessary before flushing the memtable, or for testing.
+   * As a result, no background flush task needs to be scheduled since all rows will be flushed.
+   */
+  def forceCommit(): Unit = {
+    while (tempRows.nonEmpty) flushRowsToChunks()
+  }
```

Comment:
Or maybe..
     
     def forceCommit(): Unit = for (n <- tempRows) flushRowsToChunks()